deps/GraphBLAS/Source/GxB_Matrix_tuple_iter.c

```diff
@@ -7,37 +7,80 @@
 // Sets iterator as depleted.
 static inline void _EmptyIterator
 (
-	GxB_MatrixTupleIter *iter   // Iterator to deplete.
+	_GxB_MatrixTupleIter *iter   // Iterator to deplete.
 ) {
-	iter->nvals = 0;
-	iter->nnz_idx = 0;
+	iter->nvals = 0 ;
+	iter->nnz_idx = 0 ;
 }
 
-// Create a new iterator
-GrB_Info GxB_MatrixTupleIter_new
+GrB_Info GxB_MatrixTupleIter_init
 (
-	GxB_MatrixTupleIter **iter,     // iterator to create
+	_GxB_MatrixTupleIter *iter,      // iterator to init
 	GrB_Matrix A                    // matrix to iterate over
 ) {
-	GB_WHERE(A, "GxB_MatrixTupleIter_new (A)") ;
-	GB_RETURN_IF_NULL_OR_FAULTY(A) ;
-
-	// make sure matrix is not bitmap or full
-	GxB_set(A, GxB_SPARSITY_CONTROL, GxB_SPARSE) ;
-
+	GB_WHERE(A, "GxB_MatrixTupleIter_init (iter, A)") ;
 	GrB_Index nrows ;
 	GrB_Matrix_nrows(&nrows, A) ;
+	GrB_Matrix_nvals(&iter->nvals, A) ;
+
+	GxB_Matrix_Option_get(A, GxB_SPARSITY_CONTROL, &iter->sparsity_type) ;
+	if(iter->sparsity_type != GxB_SPARSE && iter->sparsity_type != GxB_HYPERSPARSE) {
+		GB_ERROR (GrB_INVALID_VALUE,
+			"Invalid sparsity type: %d", iter->sparsity_type) ;
+	}
+	iter->A = A ;
+	iter->nnz_idx = 0 ;
+	iter->row_idx = 0 ;
+	iter->nrows = nrows ;
+	iter->p = A->p[0] ;
+	if(iter->sparsity_type == GxB_HYPERSPARSE) {
+		iter->sparse_row_idx = 0 ;
+		iter->h_size = A->nvec ;
+	}
 
+	return GrB_SUCCESS ;
+}
+
+// Create a new iterator
+GrB_Info GxB_MatrixTupleIter_new
+(
+	GxB_MatrixTupleIter **iter,     // list of iterators to create
```

Comment:
Caller shouldn't be aware that this iterator aggregates multiple nested ones
`// iterator to create`